src/main/java/net/neoforged/neoforge/transfer/resources/ResourceStack.java

```diff
@@ -0,0 +1,217 @@
+        if (amount == this.amount) return this;
+        if (amount == 0 || resource.isEmpty()) return resource().getEmptyResourceStackInstance();
+        return ResourceStack.of(resource, amount);
+    }
+
+    /**
+     * @param amount Amount to shrink by. Must be non-negative.
+     * @return A new immutable instance of this resource stack with an updated amount decreased by the specified {@code amount}.
+     *         If the newAmount is 0 or the resource is empty, then the EMPTY instance for the resource will be returned.
+     * @see #withAmount(int)
+     */
+    public ResourceStack<T> shrink(int amount) {
+        TransferPreconditions.checkNonNegative(amount);
+        return withAmount(Math.max(this.amount - amount, 0));
+    }
+
+    /**
+     * @param amount Amount to grow by. Must be non-negative.
+     *
+     * @return A new immutable instance of this resource stack with an amount increased by the specified {@code amount}.
+     *         If the resource was already empty, then the EMPTY instance will be returned instead
+     */
+    public ResourceStack<T> grow(int amount) {
+        TransferPreconditions.checkNonNegative(amount);
+        return withAmount(this.amount + amount);
+    }
+
+    /**
+     * @return A new this instance with an updated resource should it have changed, otherwise it returns itself.
+     */
+    public ResourceStack<T> with(UnaryOperator<T> operator) {
+        return with(operator, amount);
+    }
+
+    /**
+     * @param amount the amount the new stack should be.
+     * @return A new this instance with an updated resource should it have changed, otherwise it returns itself.
+     */
+    public ResourceStack<T> with(UnaryOperator<T> operator, int amount) {
+        if (amount == 0) return resource().getEmptyResourceStackInstance();
+
+        T result = operator.apply(resource);
+        if (result.equals(resource) && amount == amount()) return this;
+        return ResourceStack.of(result, amount);
+    }
+
+    /**
+     * Checks if this is empty, meaning that the amount is zero
+     * or that the resource is {@link IResource#isEmpty() empty}.
+     *
+     * @return {@code true} if empty
+     */
+    public boolean isEmpty() {
+        return ResourceHandlerUtil.isEmpty(resource(), amount());
+    }
+
+    /**
+     * Determines if the resource's backing value is enabled. If it is not, most actions are expected to not function,
+     * such as right click behaviour or similar. This is a per level/world setting. For instance, this should be checked before
+     * filling an item based resource handler.
+     * 
+     * @param enabledFeatures Feature flags currently enabled for this level.
+     * @return {@code true} if the resource is enabled in the feature set of the level; {@code false} otherwise.
+     */
+    public boolean isEnabled(FeatureFlagSet enabledFeatures) {
+        return !(resource() instanceof IRegisteredResource<?, ?> reg) || reg.isEnabled(enabledFeatures);
```

Comment:
The logic here might be slightly incorrect, iirc `ItemStack` also short circuits to being enabled when `amount <= 0` [due to it invoking `ItemStack#isEmpty`], adding a `isEmpty() ||` to the beginning should fix this.